crates/mun_language_server/src/completion/context.rs

```diff
@@ -0,0 +1,138 @@
+use crate::db::AnalysisDatabase;
+use crate::FilePosition;
+use hir::semantics::{Semantics, SemanticsScope};
+use hir::AstDatabase;
+use mun_syntax::{ast, utils::find_node_at_offset, AstNode, SyntaxNode, TextRange, TextSize};
+use ra_ap_text_edit::Indel;
+
+/// A `CompletionContext` is created to figure out where exactly the cursor is.
+pub(super) struct CompletionContext<'a> {
+    pub sema: Semantics<'a>,
+    pub scope: SemanticsScope<'a>,
+    pub db: &'a AnalysisDatabase,
+    pub position: FilePosition,
+
+    /// True if the context is currently at a trivial path.
+    pub is_trivial_path: bool,
+
+    /// True if the context is currently on a parameter
+    pub is_param: bool,
+
+    /// True if we're at an ast::PathType
+    pub is_path_type: bool,
+
+    /// The receiver if this is a field or method access, i.e. writing something.$0
+    pub dot_receiver: Option<ast::Expr>,
+}
+
+impl<'a> CompletionContext<'a> {
+    /// Tries to construct a new `CompletionContext` with the given database and file position.
+    pub fn new(db: &'a AnalysisDatabase, position: FilePosition) -> Option<Self> {
+        let sema = Semantics::new(db);
+
+        let original_file = sema.parse(position.file_id);
+
+        // Insert a fake identifier to get a valid parse tree. This tree will be used to determine
+        // context. The actual original_file will be used for completion.
+        let file_with_fake_ident = {
+            let parse = db.parse(position.file_id);
+            let edit = Indel::insert(position.offset, String::from("intellijRulezz"));
```

Comment:
Mun Rulezz!

Reply:
"intellijRulezz" is a standard string used for this purpose